### DISCUSSION TITLE
Lab 4: MCycle.v will benefit from using Funct3 over MCycleOp

### DISCUSSION THREAD

**Discussion by kxing28** [NUS-CG3207/labs#24] [Lab 4] [answered]
![image](https://github.com/user-attachments/assets/e762cc33-4b3a-4fa2-929d-591b3b72039f)

We can see above that every MCycle operation is defined by `Funct3`. Currently, we use Decoder.v to convert `Funct3` into `MCycleOp`. 

For Lab 3, this was sufficient. But for Lab 4, we need to implement `MULHSU`, so Mcycle.v needs to sign-extend one `Operand1` and keep the other `Operand2` the same. `McycleOp` is only 2 bits long so it doesn't contain the info necessary to make the sign-extension decision.

We're worried about changing structure/arguments of modules so we just wanted to make sure: Would it be alright to remove `MCycleOp` as an argument for Mcycle.v and to replace it with `Funct3`?

**Comment by r-panicker** [comment, marked as answer]
It is ok to make changes as appropriate for Lab 4. Having said that, many of the control signals can be derived similarly by choosing appropriate bits. The RISC-V decoder doesn't need to do as much decoding as is typical for other processors. Keeping the decoder has its merits - even if it is just renaming signals, it makes it comprehensible. The renaming does not add to the complexity much, as realizing they are essentially the same signal is among the easiest things the synthesis tool can do :)
On a related note, we removed `mulhsu` from the basic requirements, as it requires a bit of extra effort. The idea is to add those missing instructions that require only minimal effort and allow most compiler-generated code to be used. As long as you don't multiply a signed by an unsigned, `mulhsu` is unlikely to be emitted. 
If you have done it already, you can show it as an enhancement.

PS: Kindly use discussions instead of issues. Missed this out - apologies :(

**Reply by kxing28** [reply]
No problem, thanks for your help!
